com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/verifier/JarVerifier.java

```diff
@@ -0,0 +1,162 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.JarURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.stream.Collectors;
+
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.ResourceAccessException;
+
+import com.google.common.base.Splitter;
+import com.sap.cloud.lm.sl.cf.core.util.FileUtils;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+import com.sap.cloud.lm.sl.common.SLException;
+
+@Component
+public class JarVerifier implements Verifier {
+
+    private static final String META_INF = "META-INF";
+    private static final String CN = "CN";
+    private static final int BUFFER_SIZE = 8 * 1024;
+
+    @Override
+    public void verify(List<X509Certificate> targetCertificates, String certificateCN, URL jarURL) {
+        try {
+            JarFile jarFile = openJarFile(jarURL);
+            List<JarEntry> jarEntries = getJarEntries(jarFile);
+            verifyJarEntries(targetCertificates, jarEntries, certificateCN);
+        } catch (IOException e) {
+            throw new ResourceAccessException(MessageFormat.format(Messages.COULD_NOT_VERIFY_ARCHIVE_SIGNATURE, e.getMessage()), e);
+        } catch (SecurityException | SLException e) {
+            throw new SLException(MessageFormat.format(Messages.COULD_NOT_VERIFY_ARCHIVE_SIGNATURE, e.getMessage()));
+        }
+    }
+
+    private JarFile openJarFile(URL jarUrl) throws IOException {
+        jarUrl = toJarUrl(jarUrl);
+        JarURLConnection connection = (JarURLConnection) jarUrl.openConnection();
+        return connection.getJarFile();
+    }
+
+    private URL toJarUrl(URL jarUrl) throws MalformedURLException {
+        return new URL("jar:" + jarUrl.toString() + "!/");
+    }
+
+    private List<JarEntry> getJarEntries(JarFile jarFile) throws IOException {
+        List<JarEntry> jarEntries = new ArrayList<>();
+        Enumeration<JarEntry> entries = jarFile.entries();
+        while (entries.hasMoreElements()) {
+            JarEntry jarEntry = readJarEntry(jarFile, entries);
+            if (!jarEntry.isDirectory()) {
+                jarEntries.add(jarEntry);
+            }
+        }
+        return jarEntries;
+    }
+
+    private JarEntry readJarEntry(JarFile jarFile, Enumeration<JarEntry> entries) throws IOException {
+        JarEntry jarEntry = entries.nextElement();
+        FileUtils.validatePath(jarEntry.getName());
+        if (jarEntry.isDirectory()) {
```

Comment:
This check is repeated here and in the method above. Find a way to skip it in one of the places.

Reply:
It could be skipped here